server/src/main/java/com/saralein/server/handler/FileHandler.java

```diff
@@ -1,39 +1,100 @@
+    public FileHandler(FileHelper fileHelper, IO fileIO) {
         this.fileHelper = fileHelper;
+        this.fileIO = fileIO;
     }
 
     @Override
-    public Response handle(Request request) throws IOException {
-        String requestMethod = request.getMethod();
-        byte[] body = requestMethod.equals(Methods.GET.name()) ? createBody(request) : new byte[]{};
+    public Response handle(Request request) throws IOException, NoSuchAlgorithmException {
+        String method = request.getMethod();
+
+        if (isHeadRequest(method)) {
+            return getHeadResponse(request);
+        }
 
+        if (isPatchRequest(method)) {
+            return getPatchResponse(request);
+        }
+
+        return getGetResponse(request);
+    }
+
+    private Response getGetResponse(Request request) throws IOException {
         return new Response.Builder()
-                    .status(200)
-                    .addHeader("Content-Type", getMimeType(request))
-                    .body(body)
-                    .build();
+                .status(200)
+                .addHeader("Content-Type", getMimeType(request))
+                .body(createBody(request))
+                .build();
+    }
+
+    private Response getHeadResponse(Request request) {
+        return new Response.Builder()
+                .status(200)
+                .addHeader("Content-Type", getMimeType(request))
+                .build();
+    }
+
+    private Response getPatchResponse(Request request) throws IOException, NoSuchAlgorithmException {
+        String requestEtag = request.getHeader("If-Match");
+        byte[] body = createBody(request);
+        String resourceEtag = generateETag(body);
+
+        if (requestEtag.equals(resourceEtag)) {
+            return getPatchSuccessResponse(request);
+        }
+
+        return getPatchFailResponse();
+    }
+
+    private Response getPatchSuccessResponse(Request request) throws IOException {
+        Path resource = fileHelper.createAbsolutePath(request.getUri());
+        fileIO.write(resource, request.getBody());
+
+        return new Response.Builder()
+                .status(204)
+                .addHeader("Content-Location", request.getUri())
+                .build();
+    }
+
+    private Response getPatchFailResponse() {
+        return new ErrorResponse(409).respond();
+    }
+
+    private boolean isHeadRequest(String method) {
+        return method.equals(Methods.HEAD.name());
+    }
+
+    private boolean isPatchRequest(String method) {
+        return method.equals(Methods.PATCH.name());
     }
 
     private byte[] createBody(Request request) throws IOException {
-        String resourceUri = fileHelper.createAbsolutePath(request.getUri());
-        Path resource = Paths.get(resourceUri);
+        Path resource = fileHelper.createAbsolutePath(request.getUri());
+
+        return fileIO.read(resource);
+    }
+
+    private String generateETag(byte[] body) throws NoSuchAlgorithmException {
```

Comment:
Generating the entity tag feels like it might be it's own responsibility

Reply:
[Moved this further up](https://github.com/saralein/java-server/pull/44/files#diff-0595a804fcef17880df6ccdd28738e35R15).  I debated making it its own class, but right now it is in the File class I broke out.

Reply:
👍 

Also, I would be cautious of allowing that exception to bubble up. It would be unfortunate if everything above it in the call stack had to now be aware of `NoSuchAlgorithmException`